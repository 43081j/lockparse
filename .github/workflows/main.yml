name: Build & Test

on:
  push:
    branches: [main]
  pull_request:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
      - name: Use Node v22.x
        uses: actions/setup-node@a0853c24544627f65ddf259abe73b1d18a591444 # v5.0.0
        with:
          node-version: 22.x
          registry-url: "https://registry.npmjs.org"
          cache: "npm"
      - name: Install Dependencies
        run: npm ci --ignore-scripts
      - name: Lint
        run: npm run lint

  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [18.x, 20.x, 22.x]
      fail-fast: false
    steps:
      - name: Checkout
        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
      - name: Use Node v${{ matrix.node-version }}
        uses: actions/setup-node@a0853c24544627f65ddf259abe73b1d18a591444 # v5.0.0
        with:
          node-version: ${{ matrix.node-version }}
          registry-url: "https://registry.npmjs.org"
          cache: "npm"
      - name: Install Dependencies
        run: npm ci --ignore-scripts
      - name: Test
        run: npm run test
