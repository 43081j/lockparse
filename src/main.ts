import type {ParsedLockFile, LockFileType, PackageJsonLike} from './types.js';
import {parseNpm} from './parsers/npm.js';
import {parseYarn} from './parsers/yarn.js';
import {parsePnpm} from './parsers/pnpm.js';
import {parseBun} from './parsers/bun.js';

const typeMap: Record<string, LockFileType> = {
  'package-lock.json': 'npm',
  'yarn.lock': 'yarn',
  'pnpm-lock.yaml': 'pnpm',
  'bun.lock': 'bun',
  npm: 'npm',
  yarn: 'yarn',
  pnpm: 'pnpm',
  bun: 'bun'
};

export * from './types.js';
export * from './traverse.js';

export function parse(
  input: string,
  typeOrFileName: string,
  packageJson?: PackageJsonLike
): Promise<ParsedLockFile> {
  const lockFileType = typeMap[typeOrFileName];

  switch (lockFileType) {
    case 'npm':
      return parseNpm(input);
    case 'yarn':
      return parseYarn(input, packageJson);
    case 'pnpm':
      return parsePnpm(input);
    case 'bun':
      return parseBun(input);
    default:
      throw new Error(`Unsupported lock file type: ${typeOrFileName}`);
  }
}
