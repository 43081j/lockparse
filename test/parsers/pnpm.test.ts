import {describe, test, expect} from 'vitest';
import {parse} from '../../src/main.js';
import path from 'node:path';
import {fileURLToPath} from 'node:url';
import {readFile} from 'node:fs/promises';

const fixtureDir = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  '../fixtures'
);

describe('pnpm parser', () => {
  test('parse simple pnpm lock file', async () => {
    const input = `
lockfileVersion: '9.0'

settings:
  autoInstallPeers: true
  excludeLinksFromLockfile: false

importers:
  .:
    dependencies:
      some-package:
        version: 1.2.3
        specifier: ^1.2.3

packages:
  'some-package@1.2.3':
    resolution:
      {
        integrity: sha512-abc123...
      }
    engines:
      node: '>=12'
`;
    const parsed = await parse(input, 'pnpm');
    expect(parsed).toMatchSnapshot();
  });

  test('parses a complex pnpm lock file', async () => {
    const input = await readFile(
      path.join(fixtureDir, 'pnpm-lock.yaml'),
      'utf-8'
    );
    const parsed = await parse(input, 'pnpm');
    expect(parsed).toMatchSnapshot();
  });
});
