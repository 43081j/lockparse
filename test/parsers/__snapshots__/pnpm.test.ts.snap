// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pnpm parser > parse simple pnpm lock file 1`] = `
{
  "packages": [
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "some-package",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "1.2.3",
    },
  ],
  "root": {
    "dependencies": [
      {
        "dependencies": [],
        "devDependencies": [],
        "name": "some-package",
        "optionalDependencies": [],
        "peerDependencies": [],
        "version": "1.2.3",
      },
    ],
    "devDependencies": [],
    "name": "root",
    "optionalDependencies": [],
    "peerDependencies": [],
    "version": "",
  },
  "type": "pnpm",
}
`;

exports[`pnpm parser > parses a complex pnpm lock file 1`] = `
{
  "packages": [
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "esbuild",
      "optionalDependencies": [
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/aix-ppc64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/android-arm",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/android-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/android-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/darwin-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/darwin-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/freebsd-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/freebsd-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-arm",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-ia32",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-loong64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-mips64el",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-ppc64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-riscv64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-s390x",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/linux-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/netbsd-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/netbsd-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/openbsd-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/openbsd-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/openharmony-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/sunos-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/win32-arm64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/win32-ia32",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
        {
          "dependencies": [],
          "devDependencies": [],
          "name": "@esbuild/win32-x64",
          "optionalDependencies": [],
          "peerDependencies": [],
          "version": "0.25.10",
        },
      ],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "picocolors",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "1.1.1",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "tinyexec",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "1.0.1",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/aix-ppc64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/android-arm",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/android-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/android-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/darwin-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/darwin-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/freebsd-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/freebsd-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-arm",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-ia32",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-loong64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-mips64el",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-ppc64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-riscv64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-s390x",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/linux-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/netbsd-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/netbsd-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/openbsd-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/openbsd-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/openharmony-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/sunos-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/win32-arm64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/win32-ia32",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
    {
      "dependencies": [],
      "devDependencies": [],
      "name": "@esbuild/win32-x64",
      "optionalDependencies": [],
      "peerDependencies": [],
      "version": "0.25.10",
    },
  ],
  "root": {
    "dependencies": [
      {
        "dependencies": [],
        "devDependencies": [],
        "name": "esbuild",
        "optionalDependencies": [
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/aix-ppc64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/android-arm",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/android-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/android-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/darwin-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/darwin-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/freebsd-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/freebsd-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-arm",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-ia32",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-loong64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-mips64el",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-ppc64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-riscv64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-s390x",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/linux-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/netbsd-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/netbsd-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/openbsd-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/openbsd-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/openharmony-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/sunos-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/win32-arm64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/win32-ia32",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
          {
            "dependencies": [],
            "devDependencies": [],
            "name": "@esbuild/win32-x64",
            "optionalDependencies": [],
            "peerDependencies": [],
            "version": "0.25.10",
          },
        ],
        "peerDependencies": [],
        "version": "0.25.10",
      },
    ],
    "devDependencies": [
      {
        "dependencies": [],
        "devDependencies": [],
        "name": "picocolors",
        "optionalDependencies": [],
        "peerDependencies": [],
        "version": "1.1.1",
      },
      {
        "dependencies": [],
        "devDependencies": [],
        "name": "tinyexec",
        "optionalDependencies": [],
        "peerDependencies": [],
        "version": "1.0.1",
      },
    ],
    "name": "root",
    "optionalDependencies": [],
    "peerDependencies": [],
    "version": "",
  },
  "type": "pnpm",
}
`;
